### DISCUSSION TITLE
Configure CDN in front of file library

### DISCUSSION THREAD

**Discussion by nikhiltri** [area17/twill#2118] [Q&A]
## Problem

We have a few files in our file library that are provided on frequently accessed pages. They would benefit from being delivered through a CDN.

However currently the [file library](https://github.com/area17/twill/blob/3.x/config/file_library.php) only allows you to specify a source to put and retrieve the file library assets, like an S3 bucket. While this is essential to its functionality, it would be great if we could also specify a CDN URL to use for asset retrieval.

## Solution

Ideally we could specify a domain to use for asset retrieval, and GET requests for file library assets will sub the CDN domain if it is present.

Is anything like this on your road map? Or have requests like this come up in the past and been handled in some way by Twill implementors?

**Comment by ifox** [comment]
Hi @nikhiltri,

This is definitely something we and other Twill users have implemented and could be better documented.

It should be as simple as creating your own file service like https://github.com/area17/twill/blob/3.x/src/Services/FileLibrary/Disk.php that you would then enable using the env variable or configuration below:

```
    'file_service' => env('FILE_LIBRARY_FILE_SERVICE', 'A17\Twill\Services\FileLibrary\Disk'),
```

In your service you can decide to return a CDN URL for a file.
